neurodamus/utils/memory.py

```diff
@@ -568,3 +577,95 @@ def validate_inputs_distribute(self, num_ranks, batch_size):
+
+    def check_all_buckets_have_gids(self, bucket_allocation, population, num_ranks, cycles):
+        """
+        Checks if all possible buckets determined by num_ranks and cycles have at least one GID
+        assigned.
+
+        Args:
+            bucket_allocation (dict): The allocation dictionary containing the assignment of GIDs
+                                      to ranks and cycles.
+            population (str): The population to check.
+            num_ranks (int): The number of ranks.
+            cycles (int): The number of cycles.
+
+        Returns:
+            bool: True if all buckets have at least one GID assigned, False otherwise.
+        """
+        rank_allocation = bucket_allocation.get(population, {})
+        for rank_id in range(num_ranks):
+            for cycle_id in range(cycles):
+                if not rank_allocation.get((rank_id, cycle_id)):
+                    return False
+        return True
+
+    @run_only_rank0
+    def distribute_cells_with_validation(self,
+                                         num_ranks,
+                                         cycles=None,
+                                         metype_file=None
+                                         ) -> Tuple[dict, dict, dict]:
+        """
+        Wrapper function to distribute cells with ever smaller assignment batches until
+        all buckets have at least one GID assigned (or fail if no valid distribution is found).
+
+        Args:
+            num_ranks (int): The number of ranks.
+            cycles (int): The number of cycles to distribute cells over.
+            metype_file (str): The path to a JSON file containing memory usage for each METype.
+
+        Returns:
+            Tuple[dict, dict, dict]: Returns the same as distribute_cells once a valid distribution
+                                     is found.
+        """
+        if not cycles:
+            cycles = 1
+        logging.info("Distributing cells across %d ranks and %d cycles", num_ranks, cycles)
+
+        valid_distribution = False
+
+        def _calculate_total_elements_per_population(self):
+            total_elements_per_population = {}
+            for population, metypes in self.pop_metype_gids.items():
+                total_elements = sum(len(gids) for gids in metypes.values())
+                total_elements_per_population[population] = total_elements
+            return total_elements_per_population
+
+        total_cells_per_population = _calculate_total_elements_per_population(self)
+        average_cells_per_population = {
+            population: total / (num_ranks * cycles)
+            for population, total in total_cells_per_population.items()}
+
+        batch_size = {
+            population: max(1, int(average / MAX_ALLOCATION_STEPS))
+            for population, average in average_cells_per_population.items()}
+
+        while not valid_distribution and any(size > 0 for size in batch_size.values()):
```

Comment:
It would be useful some description about this distribution loop logic as it's getting complex.
I understand that we try the distribution with increasingly smaller batches per population until the distribution is valid. 
But that means in the bad case it will be slower as we will make N distributions, when the Nth distribution is with a single element per batch. Also it will redo the work for all populations, even if some distributed correctly.
I thought we wanted to avoid the loop and make a single distribution with the adjusted batch size, where the last batch contains the remainder elements. Example with 92 cells and 1 rank we should get batch_size=10, where the last batch will be partially filled with the remaining two cells. Are there any constraints preventing us doing so?

Reply:
Thank you for your feedback!

As you said, the loop attempts to distribute cells with decreasing batch sizes until a valid distribution is found, which can be inefficient.

Your suggestion to perform a single distribution attempt with an adjusted batch size, where the last batch contains the remaining elements, makes sense and could improve efficiency. I have massively simplified the `distribute_cells_with_validation` function to implement this approach. The new implementation calculates the batch size based on the average number of cells per population and performs a single distribution attempt. If the distribution is not valid, it raises an error, prompting the user to adjust the number of ranks or cycles.

This approach should reduce the number of distribution attempts and improve overall performance. Please review the revised implementation and let me know if you have any further suggestions or concerns.